### PR TITLE
BUILD-3093 Setup a GitHub action to enforce pre-commit at CI level

### PR DESCRIPTION
**Pull request by sebastienvermeille**
# BUILD-3093 Setup a GitHub action to enforce pre-commit at CI level

## Changes

- [x] Setup a new GitHub action responsible to evaluate pre-commit-config.yaml files with pre-commit and report any rule violation.
  That way it is possible to prevent merge of PRs not passing such requirements
- [x] Document usage in README.md
- [x] Setup and Write IT tests
- [x] Integrate SonarCloud analysis
- [x] Link some working example PRs

## How was it tested ?
* some it-tests within this repository are enabled https://github.com/SonarSource/gh-action_pre-commit/actions/workflows/it-test.yml
* an internal project using it: https://github.com/SonarSource/re-ci-images/pull/476

### CODE REVIEW

File: .github/workflows/pre-commit.yml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+on:
+  pull_request:
+
+jobs:
+  pre-commit:
+    name: "pre-commit"
+    runs-on: ubuntu-latest
+    steps:
+      - uses: SonarSource/gh-action_pre-commit@feature/svermeille/BUILD-3093-Provide-a-reusable-way-at-CI-level-to-enforce-pre-commits-in-the-organization #TODO: ref tag 0.0.1 after merge
```
Comment (julien-carsique-sonarsource) [inline .github/workflows/pre-commit.yml:9]: A trick seems to be:
- create a v1 branch equal to `feature/svermeille/BUILD-3093-Provide-a-reusable-way-at-CI-level-to-enforce-pre-commits-in-the-organization`
- use v1 ref here
- after merge, create a 1.0.0 tag
- after merge, update the v1 ref to the new tag

File: README.md [hunk 2]
```diff
@@ -1,2 +1,56 @@
...
+
+## Versioning
+
+This project is using [Semantic Versioning](https://semver.org/).
+
+The `master` branch shall not be referenced by end-users,
+please use tags instead and Renovate to stay up to date.
+
+See [GitHub: Using tags for release management](https://docs.github.com/en/actions/creating-actions/about-custom-actions#using-tags-for-release-management)
+and [GitHub: Keeping your actions up to date with Dependabot](https://docs.github.com/en/code-security/supply-chain-security/keeping-your-dependencies-updated-automatically/keeping-your-actions-up-to-date-with-dependabot)
```
Comment (julien-carsique-sonarsource) [inline README.md:52]: We rather recommend Renovate.

**Review by julien-carsique-sonarsource — APPROVED** [review]

File: .github/workflows/pre-commit.yml (hunk 1 above)
Comment (sebastienvermeille) [reply .github/workflows/pre-commit.yml:9]: I don't plan to use any v* branch for this project as this leads to more problems and maintenance efforts than it solves.
I will change this to a tagged release when this gets merged and let renovate keep it up to date :+1:

File: README.md (hunk 2 above)
Comment (sebastienvermeille) [reply README.md:52]: This doc is inspired from gh-action_release https://github.com/SonarSource/gh-action_release/blob/master/README.md#versioning

I will provide a link to both tools however. Good idea

File: README.md [hunk 3]
```diff
@@ -1,2 +1,56 @@
...
+
+jobs:
+  pre-commit:
+    name: "pre-commit"
+    runs-on: ubuntu-latest
+    steps:
+      - uses: SonarSource/gh-action_pre-commit@0.0.1 <--- replace with last tag
+        with:
+          extra-args: --from-ref=origin/${{ github.event.pull_request.base.ref }} \
+            --to-ref=${{ github.event.pull_request.head.sha }}
```
[the remark this answers is not in the document]
Comment (sebastienvermeille) [reply README.md:30]: From an OSS perspective, I think this github-action should not be too opinionated giving some default behaviours that would not work if used on a master branch. (no diff to eval if configured with `--from-ref --to-ref`)

I will document a workflow for a master branch in the README.md

Also if we take pre-commit, the philosophy used by its authors: `pre-commit run` does not check about changes regarding another git branch. This is achieved via some explicit options.

Does that make sense ?

**Pull request merged**